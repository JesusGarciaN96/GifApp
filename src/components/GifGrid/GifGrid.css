.gif-grid {
  display: flex;
  align-items: center;
  justify-content: space-around;
  flex-wrap: wrap;
  margin-bottom: 50px;
}

.gif-grid__contenedor {
  width: 300px;
  height: 250px;
  border: solid 1px #323334;
  border-radius: 5px;
  padding: 5px;
  margin: 10px;
  overflow: hidden;
}

.gif-grid__titulo {
  text-align: center;
}

.gif-grid__gif {
  max-width: 300px;
  max-height: 230px;
  margin: 0 auto;
}
