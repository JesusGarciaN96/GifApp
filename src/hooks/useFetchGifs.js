import { useState, useEffect } from 'react';

import { getGifs } from '../utils/getGifs';

export const useFetchGifs = category => {
  const [initialState, setInitialState] = useState({
    data: [],
    loading: true,
  });

  useEffect(() => {
    (async () => {
      try {
        const data = await getGifs(category);
        setInitialState({
          data,
          loading: false,
        });
      } catch (ex) {
        console.log(ex);
      }
    })();
  }, []);
  return initialState;
};
