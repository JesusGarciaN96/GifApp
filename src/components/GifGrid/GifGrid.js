import React, { useState, useEffect, Fragment } from 'react';
import PropTypes from 'prop-types';

import './GifGrid.css';

function GifGrid({ category }) {
  const [gifCollection, setGifCollection] = useState([]);
  const apiKey = 'BQhSthjbP9NOXgxUL6W5PsjbtuKfXEa8';

  const getGifs = async () => {
    if (category) {
      try {
        const { data } = await (
          await fetch(
            `https://api.giphy.com/v1/gifs/search?api_key=${apiKey}&q=${category}&limit=10`
          )
        ).json();

        const collection = data.map(({ id, title, images }) => ({
          id,
          title,
          urlGif: images.downsized.url,
        }));
        setGifCollection(collection);
      } catch (ex) {
        console.log(ex);
      }
    }
  };

  useEffect(() => {
    getGifs();
  }, []);

  return (
    <Fragment className='gif-grid'>
      <h2 className='gif-grid__title'>{category}</h2>
      <div className='gif-grid__grid'>
        {gifCollection.map(gif => (
          <div className='gif-grid__container' key={gif.id}>
            <h5 className='gif-grid__gif-title'>{gif.title.toUpperCase()}</h5>
            <img className='gif-grid__gif' src={gif.urlGif} alt={gif.title} />
          </div>
        ))}
      </div>
    </Fragment>
  );
}

GifGrid.defaultProps = {
  category: '',
};

GifGrid.propTypes = {
  category: PropTypes.string,
};

export default GifGrid;
