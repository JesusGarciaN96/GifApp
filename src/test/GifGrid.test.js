import React from 'react';
import { shallow } from 'enzyme';
import '@testing-library/jest-dom';

import GifGrid from '../components/GifGrid/GifGrid';
import { useFetchGifs } from '../hooks/useFetchGifs';

jest.mock('../hooks/useFetchGifs');

// Se usa el mock en todos los test antes de usar el shallow
// de otra manera habra un error en la desestructuraciÃ³n de GifGrid
describe('Testing de componentes con Enzyme', () => {
  const categoria = 'One Punch Man';
  const claseContenedor = 'animate__slideInUp';

  test('Debe mostrar (renderizar) <GifGrid /> de manera correcta', () => {
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true,
    });
    const componente = shallow(<GifGrid category={categoria} />);
    expect(componente).toMatchSnapshot();
  });

  test('La categoria debe ser una cadena valida', () => {
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true,
    });
    const wrapper = shallow(<GifGrid category={categoria} />);
    expect(wrapper.find('.gif-grid__title').text().trim()).toBe(categoria);
  });

  test('Verificar que la clase sea correcta', () => {
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true,
    });
    const wrapper = shallow(<GifGrid category={categoria} />);
    expect(wrapper.find('.gif-grid').prop('className').split(' ')).toContain(
      claseContenedor
    );
  });

  test('Validar que la url sea de una imagen', () => {
    const fakeDataResponse = [
      {
        id: 1,
        title: 'Saitama',
        urlGif: 'https://localhost/some/saitama.png',
      },
    ];
    useFetchGifs.mockReturnValue({
      data: fakeDataResponse,
      loading: false,
    });
    const wrapper = shallow(<GifGrid category={categoria} />);
    console.log(wrapper.find('img').at(0));
  });

  // Llama el hook de fetch y usa un mock para simular que si respondio
  // y verificamos que coincida con el sanpshot
  test('Debe mostrar items cuando se carga el componente', () => {
    const fakeDataResponse = [
      {
        id: 1,
        title: 'Saitama',
        urlGif: 'https://localhost/some/saitama.png',
      },
    ];
    useFetchGifs.mockReturnValue({
      data: fakeDataResponse,
      loading: false,
    });
    const componente = shallow(<GifGrid category={categoria} />);
    expect(componente).toMatchSnapshot();
  });
});
