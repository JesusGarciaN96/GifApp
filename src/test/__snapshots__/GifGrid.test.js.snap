// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing de componentes con Enzyme Debe mostrar (renderizar) <GifGrid /> de manera correcta 1`] = `
<Fragment>
  <h2>
    Cargando Im√°genes...
  </h2>
  <div
    className="gif-grid animate__animated animate__slideInUp"
  >
    <h2
      className="gif-grid__title"
    >
      One Punch Man
    </h2>
    <div
      className="gif-grid__grid"
    />
  </div>
</Fragment>
`;

exports[`Testing de componentes con Enzyme Debe mostrar items cuando se carga el componente 1`] = `
<Fragment>
  <div
    className="gif-grid animate__animated animate__slideInUp"
  >
    <h2
      className="gif-grid__title"
    >
      One Punch Man
    </h2>
    <div
      className="gif-grid__grid"
    >
      <div
        className="gif-grid__container"
        key="1"
      >
        <h5
          className="gif-grid__gif-title"
        >
          SAITAMA
        </h5>
        <img
          alt="Saitama"
          className="gif-grid__gif"
          src="https://localhost/some/saitama.png"
        />
      </div>
    </div>
  </div>
</Fragment>
`;
